### ISSUE TITLE
Optimization variable selector for flash

### ISSUE THREAD

**Issue by stepanzh** [vvpisarev/CubicEoS.jl#18]
The corresponding branch is `variable_dispatch`, I made it from `dev`.

The aim of the dispatch is a comfort way to switch between different optimization variables.

I see them in a manner of

```julia
"See [`PhysicalVariables`](@ref), [`FractionVariables`](@ref), [`IdealIdentityVariables`](@ref)."
abstract type VTFlashStateVariables end

"Moles and volume."
struct PhysicalVariables <: VTFlashStateVariables
end

"Mole of i-th component in 1st phase by total moles of i-th component, saturation."
struct FractionVariables <: VTFlashStateVariables
end

"Variables for which ideal part of hessian of compositions is identity matrix."
struct IdealIdentityVariables <: VTFlashStateVariables
end

# Dispatch
vt_flash(vars::PhysicalVariables, mix, nmol, volume, RT) = ...
vt_flash(vars::FractionVariables, mix, nmol, volume, RT) = ...
vt_flash(vars::IdealIdentityVariables, mix, nmol, volume, RT) = ...

# End-user API
vt_flash(mix, nmol, volume, RT; state_variables=PhysicalVariables())
```

Where the structs `<: VTFlashStateVariables` are singletons for dispatch.

@vvpisarev, you are welcome to discuss interface and implementation patterns.

**Comment by stepanzh** [comment]
162388bc55221166145111840d37f5cde9bac30e relates to calculation of gradient and hessian for NVT statement in "physical variables": moles [mol] and volume [m3]. This calculation is "tolerant" to the singletons (see issue message).

I guess, for other sets of variables corresponding hessians may be derived by scaling from those general functions. Perhaps, the scaling may be computationally inefficient in some cases.

**Comment by stepanzh** [comment]
I also thought about some wrapping of `state` minimization-vector, e.g.

```julia
struct PhysicalVariablesState{T}
     x::Vector{T}
end

PhysicalVariablesState(mix) = ... # allocates `ncomponents(mix) + 1` vector

state = PhysicalVariablesState(mix)
```

And dispatching on type of `state`.

But this cause `.`-boilerplate like `nmol = @view state.x[1:end-1]` instead of `nmol = @view state[1:end-1]` like in current code.

Also, when we work with `VarsFoo` and want to (reuse code)
1. Call a method corresponding to `VarsBar`
2. Post-process the output, transforming to `VarsFoo`

It seems to causing extra wrapping (possibly, allocations or "remind where the pointer to `x` is and where `x` is origin vector") or some manipulation with `convert`.

**Comment by stepanzh** [comment]
# Draft of general interface

4aa9bc29bd5734a7447d90c447f6b9edb27e5a56 introduces draft of general `vt_flash` interface. For now, both versions can be ran.

The interface is on top of `src/vt_flash/vt_flash.jl` file.

`PhysicalState` -base calculation is tested on a two-phase point for (methane + n-pentane) using the draft.

**Comment by stepanzh** [comment]
# Ratio variables

875e99b4bee1a2ed9b5650e35c0798818c6b7d1f and 4b5d2b0e9bce037a9aab841d3de87dab86bebd0f introduce `RatioState` variables which are `N'1 / N1, ..., N'n / Nn, V' / V`.

These variables are still used in `vt_flash(mix, nmol, volume, RT)` with another (old) implementation. So old implementation may be compared to `vt_flash(mix, nmol, volume, RT, RatioState)`.

## Test example

For a four-component mixture the implementations give

```julia
julia> mix
BrusilovskyEoSMixture{Float64}(nitrogen + methane + propane + n-decane)

julia> nmol
4-element Vector{Float64}:
 0.0004925999999999999
 0.0004416
 0.0004416
 0.0006242

julia> volume
1.0e-6

julia> RT
4157.23130907662

julia> ratiostate = vt_flash(mix, nmol, volume, RT, CubicEoS.RatioState; gtol=1e-3/RT);

julia> fracold = vt_flash(mix, nmol, volume, RT);

julia> dump(ratiostate)
CubicEoS.VTFlashResult{Float64, CubicEoS.RatioState{Vector{Float64}}}
  singlephase: Bool false
  RT: Float64 4157.23130907662
  nmolgas: Array{Float64}((4,)) [0.0004516319327931407, 0.0003890882771558721, 0.0003401765215206347, 0.00013229574796212154]
  volumegas: Float64 8.378179547180461e-7
  nmolliq: Array{Float64}((4,)) [4.096806720685924e-5, 5.251172284412791e-5, 0.00010142347847936527, 0.0004919042520378785]
  volumeliq: Float64 1.6218204528195384e-7
  state: CubicEoS.RatioState{Vector{Float64}}
    x: Array{Float64}((5,)) [0.08316700610405856, 0.1189124158607969, 0.22967273206378006, 0.7880555143189337, 0.16218204528195385]
  optim: CubicEoS.VTFlashOptimStats
    converged: Bool true
    iters: Int64 7
    calls: Int64 8

julia> dump(fracold)
CubicEoS.VTFlashResult{Float64, Vector{Float64}}
  singlephase: Bool false
  RT: Float64 4157.23130907662
  nmolgas: Array{Float64}((4,)) [0.00045163193279314093, 0.0003890882771558725, 0.0003401765215206362, 0.00013229574796212186]
  volumegas: Float64 8.378179547180467e-7
  nmolliq: Array{Float64}((4,)) [4.096806720685899e-5, 5.2511722844127456e-5, 0.00010142347847936376, 0.0004919042520378781]
  volumeliq: Float64 1.6218204528195325e-7
  state: Array{Float64}((5,)) [0.08316700610405806, 0.11891241586079587, 0.22967273206377664, 0.7880555143189332, 0.16218204528195326]
  optim: CubicEoS.VTFlashOptimStats
    converged: Bool true
    iters: Int64 7
    calls: Int64 8

julia> norm(ratiostate.state.x - fracold.state)
3.688777608244322e-15
```

Same results!

P.S. for now I'm considering
- refactor for closure generation, because they seem very similar for `RatioState` and `PhysicalState`;
- a destructive option for `nmolvol(s::AbstractVTFlashState...)` for memory saving;
- non-allocating versions for `gradient!` and `hessian!(hess, s::AbstractVTFlashState, ...)`.

**Comment by stepanzh** [comment]
# Ideal Identity variables

1e7bd57936849834eb0cbc1d79451f5bce901ab9 and 310c0be0dc759f1282d7b193d8316cbd0da1b10d implement `IdealIdentityState` for which ideal part of hessian over moles is identity matrix.

```
    IdealIdentityState

Variables that make ideal part of hessian over moles to identity matrix.

state = [ 2 √Nᵢ arcsin(√N'ᵢ / √Nᵢ); √N V'/V ], where N = Σᵢ Nᵢ (total moles).
         |-----------------------| |-------|
                    α'ᵢ               α'V
```

**Comment by stepanzh** [comment]
4dfe4a81d0dc14efc4bf0fbb46eeb704eb51528f adds comparing test between all implemented `AbstractVTFlashState`s (physical, ratio and idealidentity). They are passed under `rtol=1e-5` for moles and volumes of gas and liquid phases, when `gtol=1e-5/RT` is used in calculations.

Currently, the comparison assumes transitivity of `≈`-operator (`x ≈ y rtol=1e-5`).

The test is passed!

**Comment by stepanzh** [comment]
# Newton vtflash

0ab2a57ed7721b263eed211fa5f0c6f4daf11ea3 add support for vt_flash calculation based on Newton optimization.

The solver is (unit-)tested 351470feecbade830d6d3aa9d5a57fe3d1d31dc7 against bfgs.

**Comment by stepanzh** [comment]
TODO: check type-stability of nvt-functions and state-related functions like hessian, gradient etc.

**Comment by stepanzh** [comment]
Closed by #22.
